lib/Pipes/PipelineManager.cpp

```diff
@@ -162,26 +156,187 @@ checkComponentsVersion(const revng::DirectoryPath &ExecutionDirectory,
+    const std::string SecondExtension = DateTime.substr(Size - 6);
+
+    auto BackupFilePath = Path.addExtension(FirstExtension)
+                            .addExtension(SecondExtension);
+
+    auto MaybeExists = BackupFilePath.exists();
+
+    if (not MaybeExists)
+      return MaybeExists.takeError();
+
+    if (not MaybeExists.get())
+      return BackupFilePath;
+  }
+}
+
+static llvm::Expected<revng::FilePath>
+createModelBackup(const revng::FilePath &ModelFile) {
+  llvm::Expected<revng::FilePath>
+    ModelBackupFile = getBackupFilePath(ModelFile);
+  if (not ModelBackupFile)
+    return ModelBackupFile.takeError();
+
+  if (auto Error = ModelFile.copyTo(ModelBackupFile.get()); !!Error)
+    return Error;
+
+  return ModelBackupFile;
+}
+
+static llvm::Error migrateModelFile(const revng::FilePath &ModelFile) {
+  TemporaryFile InputTemporaryFile("revng-model-migration-input", "yml");
+  TemporaryFile OutputTemporaryFile("revng-model-migration-output", "yml");
+
+  revng::FilePath
+    InputFile = revng::FilePath::fromLocalStorage(InputTemporaryFile.path());
+  if (auto Error = ModelFile.copyTo(InputFile); !!Error)
+    return Error;
+
+  {
+    int ReturnCode = ::Runner.run("revng",
+                                  { "model",
+                                    "migrate",
+                                    InputTemporaryFile.path().str(),
+                                    "--output",
+                                    OutputTemporaryFile.path().str() });
+
+    if (ReturnCode != 0)
+      return revng::createError("Error encountered while running migrations, "
+                                "process returned: %d",
+                                ReturnCode);
+  }
+
+  revng::FilePath
+    OutputFile = revng::FilePath::fromLocalStorage(OutputTemporaryFile.path());
+  if (auto Error = OutputFile.copyTo(ModelFile); !!Error)
+    return Error;
+
+  return llvm::Error::success();
+}
+
+// Conveninece function for setUpPipeline, creates a backup for the model file
+// and runs migrations on it. On success returns the backup path, on failure
+// returns the error and restores the model file from backup. If something goes
+// horribly wrong and it fails to restore, it aborts.
+static llvm::Expected<revng::FilePath>
+migrateModel(const revng::DirectoryPath &ExecutionDirectory) {
+  auto ModelFile = ExecutionDirectory.getDirectory("context")
+                     .getFile(revng::ModelGlobalName);
+
+  llvm::Expected<bool> ModelFileExists = ModelFile.exists();
+  if (not ModelFileExists)
+    return ModelFileExists.takeError();
+
+  if (not ModelFileExists.get()) {
+    // Model file does not exist, so no migrations will actually run (empty
+    // model is assumed to be in the latest version). So for the sake of better
+    // error reporting, process this case explicitly.
+    return revng::createError("Migrations will not be run since the model file "
+                              "does not exist");
+  }
```

Comment:
I think this should not happen. We should abort here and/or catch this situation *before* attempting a migration.